### ISSUE TITLE
Work on MRs from different projects in parallel

### ISSUE THREAD

**Issue by okainov** [smarkets/marge-bot#224]
Would be great to be able to work asynchronously with MRs from different projects

**Comment by nirizr** [comment]
I think my PR #293 might be helpful for this

**Comment by okainov** [comment]
@nirizr at least from the description I'm not sure it does what expected. Within the single project I see it as making sense to merge chronologically, this issue was more about multiple independent projects...

**Comment by nirizr** [comment]
I'm not sure I follow. Within a single project you say you rather Marge bot wait for a single MR until it passes CI although other MRs are already ready to be merged? I'm curious, why wouldn't you want a ready MR to be merged?

**Comment by okainov** [comment]
> Within a single project you say you rather Marge bot wait for a single MR until it passes CI although other MRs are already ready to be merged? I'm curious, why wouldn't you want a ready MR to be merged?

I'm not against it, it might be useful feature :) But in our team we do use rebase-only approach and it's better to have defined order of automatic merges than "fight who is faster to pass ci".

The ideal thought for a single project was that whatever merge tool does, it might build some kind of "merge train" inside itself to avoid multiple rebuilds. So if I have 3 ready-to-merge MRs, ideally it should build "virtual" branch which has all three merged and use this in order to validate the 3rd PR.

**Comment by nirizr** [comment]
I was only asking out of curiosity, it's also perfectly fine if you're against this as a feature :)

And I see why in a rebase-only flow you'd find this an inconvenience. Thanks!

**Comment by snim2** [comment]
@laurence-hudson-mindfoundry In our instance of `marge-bot`, the project regexp is just `.*` and we don't create a new instance of `marge-bot` for new groups or projects, we just add the `marge-bot` user to the group/project with `maintainer` privileges. 

This does mean that long-running CI pipelines can "hold up" other MRs, but it does keep everything very simple.

**Comment by laurence-hudson-mindfoundry** [comment]
> @laurence-hudson-mindfoundry In our instance of `marge-bot`, the project regexp is just `.*` and we don't create a new instance of `marge-bot` for new groups or projects, we just add the `marge-bot` user to the group/project with `maintainer` privileges.
> 
> This does mean that long-running CI pipelines can "hold up" other MRs, but it does keep everything very simple.

Yeah we went down that route for a while, but the delays introduced where just too high, hence the fork.

**Comment by snim2** [comment]
> Yeah we went down that route for a while, but the delays introduced where just too high, hence the fork.

Right, I feel your pain! I can't remember whether Marge sets [merge when pipeline succeeds](https://docs.gitlab.com/ee/user/project/merge_requests/merge_when_pipeline_succeeds.html) after it's pushed a rebased version of each MR, but that might speed things up a little.

**Comment by nicolas17** [comment]
I can't believe it doesn't already work that way, if I hadn't found this issue I wouldn't have known until trying marge-bot in production :neutral_face:

You're saying it will currently wait for CI to finish in one MR, merge it, and *then* rebase the next MR in the queue *even if they are on different projects*?
